### DISCUSSION TITLE
Request: Firebase Authentication emulator support

### DISCUSSION THREAD

**Discussion by nicoburns** [invertase/react-native-firebase#4461] [Ideas] [answered]
### Background
Firebase have just released an emulator for Firebase Authentication as part of the Firebase Emulator Suite, and the latest version of the Android and iOS SDKs support it:

- Android release notes: https://firebase.google.com/support/release-notes/android#auth_v20-0-0
- iOS release notes: https://firebase.google.com/support/release-notes/ios#version_700_-_october_26_2020
- Usage documentation: https://firebase.google.com/docs/emulator-suite/connect_auth

### Feature Request
I would like to request that this is implemented in React Native Firebase. Additionally, while there are docs for firestore [here](https://rnfirebase.io/firestore/emulator), I couldn't find any documentation on using the Realtime Database emulator with React Native Firebase. It would be good if that could be documented (and implemented if it's not).

### Motivation / Use Case

This is a big deal for us as it enables:

- A completely local isolated / offline dev workflow. We're currently using a shared hosted development instance which works well enough, but is far from ideal.
- Spinning up a complete version of our stack for e2e testing in our CI pipelines.

**Comment by mikehardy** [comment, marked as answer]
This is actually a semi-urgent need for us here for maintenance reasons, our E2E CI flow frequently fails because auth API requests are rate-limited. This is on the roadmap but it's just a hair behind forward-porting to the just-released firebase-ios-sdk v7.0.0 breaking changes. That priority order is because there are currently outstanding upstream issues that should be resolved with fresh upstream releases, but of course only if we can access them. 

If you were to pick this up and propose a PR for it though I would collaborate as much as needed to help you land it, then it could be as fast as you could implement instead of waiting on limited maintainer time here. The tests directory here has a README that helps get set up and there are examples (both in the packages and in the E2E tests) of how to use other APIs via emulator (specifically firestore recently, also functions)

In the absence of community contribution no ETA can be committed by I personally hope sometime in a 4-6 week time frame.

**Reply by nicoburns** [reply]
If it were possible to implement this standalone then I'd be tempted to have a go. Alas, I think this feature is also blocked on firebase-ios-sdk v7.0.0 (and android authentication v20.0.0). Maybe I can help with the docs once it's implemented.

**Reply by mikehardy** [reply]
forward-port to firebase-ios-sdk v7.0.0 is second priority with ETA of probably 1-2 weeks then, first I need to peel the machine learning APIs that were split from firebase to 'Google MLKit' out to a standalone as they are not forward-portable (ETA this week), then I can do the port. All ETAs subject to change of course but those are my concrete tasks

**Reply by mikehardy** [reply]
v10, just released, has auth emulator support! Enjoy

**Reply by leblancmeneses** [reply]
Is there a template that has the local emulator working? I'm trying to identify where I went wrong.  The typings in vscode shows intellisense for ```firebase.auth().useEmulator(url: string)```, however, I get  ``` ERROR    TypeError: _app.default.auth().useEmulator is not a function.``` at runtime.

```
    "@react-native-firebase/app": "^10.6.4",
    "@react-native-firebase/auth": "^10.6.4",
    "@react-native-firebase/crashlytics": "^10.6.4",
    "@react-native-firebase/database": "^10.6.4",
    "@react-native-firebase/firestore": "^10.6.4",
    "@react-native-firebase/messaging": "^10.6.4",
    "@react-native-firebase/storage": "^10.1.1",

import firebase from '@react-native-firebase/app';

// ${host} evaluates to 192.168.157.2

if (__DEV__) {
  const matches = env.siteOrigin.match(/.*\/\/([^:]+)/);
  if (matches) {
    const host = matches[1];
    firebase.auth().useEmulator(`http://${host}:9099`);
    firebase.firestore().settings({
      host: `${host}:8080`,
      ssl: false,
    });
  }
}
```

**Reply by mikehardy** [reply]
Don't mix and match versions, this is bad:

```
    "@react-native-firebase/storage": "^10.1.1",
```

If useEmulator is not a function then you have something stale somewhere either caching `npx react-native-clean-project` or you have overridden the native SDKs and are not on ios sdk v7+ / android sdk v26+ from firebase.

There is always a template for everything react-native-firebase does, it's in the packages/**/__tests__, packages/**/e2e test implementations and packages/tests e2e test app. ergo, https://github.com/invertase/react-native-firebase/blob/c832407fbb415f5400172df7c3d75e81fba0032c/tests/app.js#L47

**Reply by leblancmeneses** [reply]
Thanks for the pro tips.  Besides what you pointed out another problem I had locally is on android google-services.json project_id must match the same value the emulator was started on```npx firebase --project=ci emulators:start```.  The mobile app was created by another team member so that caused some initial hiccups.  Thanks.

**Reply by mikehardy** [reply]
If you're just getting up to speed and your first thing is a react-native-firebase project it is indeed a *lot* to get up to speed on. Hopefully you're moving forward now - good luck on your project

**Reply by dooleyb1** [reply]
@mikehardy do these emulators work for [Phone Auth](https://rnfirebase.io/auth/phone-auth)? I've somewhat of an implementation in place but seem to be getting the following error so just want to make sure this is supported before I go digging further as I couldn't find any documentation:

```
NativeFirebaseError: [auth/invalid-verification-code] The SMS verification code used to create the phone auth credential is invalid. Please resend the verification code SMS and be sure to use the verification code provided by the user.
```

The auth emulator running locally is showing the phone number verification code and also creating a user in the `Authentication` section of my emulator UI however the app doesn't seem to like the verification code.

**Reply by mikehardy** [reply]
Browse into our code repo, packages/auth/e2e and find the phone auth parts. You have to grab the oob code from the emu but it definitely works, I wrote that myself and see it working near daily. Sorry no links I'm on mobile at the moment and it's much hard to dig them out

**Reply by dooleyb1** [reply]
Cheers @mikehardy I was actually using the nice helper you made in the E2E tests to fetch the code but our project doesn't seem to like it. Most likely an error on our side then configuration wise. Will keep digging, thanks 👍
